src/main/java/org/spongepowered/common/item/inventory/util/ContainerUtil.java

```diff
@@ -200,18 +219,92 @@ public static void performBlockInventoryDrops(WorldServer worldServer, double x,
         }
     }
 
-    // TODO Inventory - Container lens
-    // TODO Inventory - Add a fallback Container lens
     @SuppressWarnings("unchecked")
     @Nullable
-    public static MinecraftLens getLens(net.minecraft.inventory.Container container, SlotCollection collection) {
-        if (container instanceof ContainerChest) {
-            return new ContainerChestInventoryLens((InventoryAdapter<IInventory, ItemStack>) container, collection, ((ContainerChest) container).numRows);
-        } else if (container instanceof ContainerPlayer) {
-            return new ContainerPlayerInventoryLens((InventoryAdapter<IInventory, ItemStack>) container, collection);
+    public static Lens<IInventory, ItemStack> getLens(Fabric<IInventory> fabric, net.minecraft.inventory.Container container, SlotCollection slots) {
+        // TODO use fabric instead if possible?
+
+        System.out.print("### getLens for: " + container.getClass().getName() + " ###\n");
+        // Container is Adapter?
+        if (container instanceof InventoryAdapter) {
+            Lens lens = ((InventoryAdapter) container).getRootLens();
+            if (lens != null) {
+                System.out.print("Adapter found\n");
+                return lens;
+            }
+            // else lens is null: Adapter is probably getting initialized
         }
 
-        return null;
+        // Container provides Lens?
+        if (container instanceof LensProvider) {
+            System.out.print("LensProvider found\n");
+            InventoryAdapter<IInventory, ItemStack> adapter;
+            if (container instanceof InventoryAdapter) {
+                adapter = ((InventoryAdapter) container);
+            } else {
+                adapter = new Adapter(MinecraftFabric.of(container));
+            }
+            return ((LensProvider) container).getRootLens(null, adapter); // TODO 1st param is null because we have a container and not IInventory
```

Comment:
`inv` parameter of `getRootLens` should be annotated with `@Nullable` if this `null` is to remain.